sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala

```diff
@@ -225,9 +227,13 @@ class ForgeMessageBuilder(mainchainSynchronizer: MainchainSynchronizer,
       if (branchPointInfo.referenceDataToInclude.size == withdrawalEpochMcBlocksLeft) { // SC block is going to become the last block of the withdrawal epoch
         Seq() // no SC Txs allowed
       } else { // SC block is in the middle of the epoch
-        nodeView.pool.take(SidechainBlock.MAX_SIDECHAIN_TXS_NUMBER) // TO DO: problems with types
+        val mempoolTransactions = nodeView.pool.take(SidechainBlock.MAX_SIDECHAIN_TXS_NUMBER) // TO DO: problems with types
           .map(t => t.asInstanceOf[SidechainTransaction[Proposition, NoncedBox[Proposition]]])
-          .toSeq
+
+        val applicationForgeTransactions = applicationForger.onForge(parentBlockInfo, blockSignPrivateKey.publicImage())
```

Comment:
Are we sure that transactions proposed by ApplicationForger are valid?
Currently, if ApplicationForger will provide transaction that is not valid or which has a spent box intersections (double spend), it will lead to an invalid SC block proposed by the Forger.
Another point of discussion is what should be included first: `mempoolTransactions` or `applicationForgeTransactions`? We should keep in mind the max block size as well.

The first solution that come to my mind is:
1) get mempoolTransactions
2) pass them as an immutable collection as an argument to `applicationForger.onForge`
3) for the applicationForgeTransactions verify them with the compatibility checker against the previously added transactions, then against the SidechainState.
If everything is ok, forge a block.
What do you think?

Reply:
I get the point and  agree.
I thought we could count on the check  at block acceptance by the other nodes, but in this way we can stop the process earlier and avoid to forge a wrong block at all